ibis/sql/postgres/udf/api.py

```diff
@@ -0,0 +1,218 @@
+    parameters: List[str] - give names to the arguments of the UDF
+
+    Returns
+    -------
+    wrapper : Callable
+        The wrapped function
+    """
+    if parameters is None:
+        parameters = ['v{}'.format(i) for i in range(len(input_types))]
+    elif len(input_types) != len(parameters):
+        raise ValueError(
+            (
+                "Length mismatch in arguments to existing_udf: "
+                "len(input_types)={}, len(parameters)={}"
+            ).format(len(input_types), len(parameters))
+        )
+
+    udf_node_fields = collections.OrderedDict([
+        (name, Arg(rlz.value(type_)))
+        for name, type_ in zip(parameters, input_types)
+    ] + [
+        (
+            'output_type',
+            lambda self, output_type=output_type: rlz.shape_like(
+                self.args, dtype=output_type
+            )
+        )
+    ])
+    udf_node_fields['resolve_name'] = lambda self: name
+
+    udf_node = create_udf_node(name, udf_node_fields)
+
+    def _translate_udf(t, expr):
+        func_obj = sa.func
+        if schema is not None:
+            func_obj = getattr(func_obj, schema)
+        func_obj = getattr(func_obj, name)
+
+        sa_args = [t.translate(arg) for arg in expr.op().args]
+
+        return func_obj(*sa_args)
+
+    add_operation(udf_node, _translate_udf)
+
+    def wrapped(*args, **kwargs):
+        node = udf_node(*args, **kwargs)
+        return node.to_expr()
+
+    return wrapped
+
+
+def func_to_udf(conn,
+                python_func,
+                in_types,
+                out_type,
+                schema=None,
+                replace=False,
+                name=None):
+    """Defines a UDF in the database
+
+    Parameters
+    ----------
+    conn: sqlalchemy engine
+    python_func: python function
+    in_types: List[DataType]
+    out_type : DataType
+    schema: str - optionally specify the schema in which to define the UDF
+    replace: bool - replace UDF in database if already exists
+    name: str - name for the UDF to be defined in database
+
+    Returns
+    -------
+    wrapper : Callable
+        The ibis UDF object as a wrapped function
+    """
+    if name is None:
+        internal_name = python_func.__name__
+    else:
+        internal_name = name
+    signature = inspect.signature(python_func)
+    parameter_names = signature.parameters.keys()
+    replace_text = ' OR REPLACE ' if replace else ''
+    schema_fragment = (schema + '.') if schema else ''
+    template = """CREATE {replace} FUNCTION
+{schema_fragment}{name}({signature})
+RETURNS {return_type}
+LANGUAGE plpythonu
+AS $$
+{func_definition}
+return {internal_name}({args})
+$$;
+"""
+
+    postgres_signature = ', '.join(
+        '{name} {type}'.format(
+            name=name,
+            type=ibis_to_postgres_str(type_),
+        )
+        for name, type_ in zip(parameter_names, in_types)
+    )
+    return_type = ibis_to_postgres_str(out_type)
```

Comment:
We should ultimately make a SQLAlchemy construct and go through _its_ compiler rather than stringifying SQLAlchemy types and formatting them in here. Eventually that might actually be a nice upstream contribution as well.

Reply:
Let's make a follow up issue for that.

Reply:
Great suggestion! I created #1890 for this.